docs/src/proposals/append-vec-storage.md

```diff
@@ -0,0 +1,374 @@
+via `owner_block_offset` + 32 + `owner_local_id` * `OWNER_ENTRY_SIZE`.
+
+Again, as the offset of each block is aligned.  A 32-byte hash entry means each
+entry is naturally 64-bit aligned without padding.
+
+Note that the existing 8-byte `write_version` field in `StoredMeta` is removed
+in this design as we now have AccountsCache as a write cache so that we don't
+need to maintain the `write_version` to enable multiple AppendVec per slot.
+
+```
++--------------------------------------+
+| Hash (32 bytes for the entire block) |
++--------------------------------------+
+| owner 1                              |
++--------------------------------------+
+| owner 2                              |
++--------------------------------------+
+|         ...                          |
++--------------------------------------+
+| owner N                              |
++--------------------------------------+
+```
+
+```
++-------------------------------------------------------------+
+| Owner (32 bytes each)                                       |
++-------------------------------------------------------------+
+| pubkey          | 32 bytes                                  |
++-------------------------------------------------------------+
+```
+
+Again, as mentioned earlier in the Header section, the Owners Block also
+offers forward compability as long as the newer version only adds new fields
+and does not change the definition of the existing fields.
+
+#### (Small) Account Data Blocks
+One small account data block contains multiple accounts' data.
+
+```
++------------------------------------+
+| Account data block 1 (compressed)  |
++------------------------------------+
+| Account data block 2 (compressed)  |
++------------------------------------+
+| Account data block 3 (compressed)  |
++------------------------------------+
+|          ...                       |
++------------------------------------+
+| Account data block N (compressed)  |
++------------------------------------+
+```
+```
++------------------------------------------+
+| Account data block (after decompression) |
++------------------------------------------+
+| Account data 1                           |
++------------------------------------------+
+| Account data 2                           |
++------------------------------------------+
+| Account data 3                           |
++------------------------------------------+
+|          ...                             |
++------------------------------------------+
+| Account data N                           |
++------------------------------------------+
+```
+
+The account data format for both small account and blob account are the same:
+```
++-------------------------------------------------------------+
+| Account data (after decompression)                          |
++-------------------------------------------------------------+
+| rent_epoch      |  0-8 bytes                                |
```

Comment:
zero _or_ 8 bytes, right?  this should probably be appended instead of prepended and `NonZeroU64`, then we get free memory layout optimization from the compiler (assuming rent-epoch == 0 is an invalid thing)

Reply:
Thanks for the correction.  Yes, it should be 0 or 8 bytes.

If we read account data more often than rent-epoch (I think this assumption should be yes), then yes it should be appended instead of prepended.

Reply:
`rent_epoch` is quite the story. It is only relevant anymore for rent-paying accounts, which is a shrinking set of ~1M accounts on mnb. For everyone else, I have a feature to set the value to Epoch::MAX for all rent exempt accounts. Thus, we don't need to store it at all for rent exempt accounts (the vast bulk of accounts). For rent paying accounts it is 1 of 4 values. Slot - 1, Slot, Slot + 1, 0. Where Slot is the slot of the append vec itself. It is possible we could reduce these values down to fewer than 4. Maybe even 2 values:
1. rent already collected as of this slot
2. rent last collected as of prior slot

Reply:
I am hustling to get this feature ready.
https://github.com/solana-labs/solana/issues/28683